### ISSUE TITLE
Make constraint solving modes individually toggleable

### ISSUE THREAD

**Issue by MxmUrw** [DiffMu/DiffPrivacyInferenceHs#39]
In order to be able to test our different constraint solving strategies, we should be able to enable/disable the solving modes individually, such that e.g. we can test subtyping without having constraints contracted by the diamond-contractor.

- [x] Implement.
- [x] Fix tests.

**Comment by MxmUrw** [comment]
During solving of the test "contracts a larger cycle with more stuff" we get
```
Debug[Constr]: constr_8 : [worst,global,exact] IsSupremum (τ_5,τ_0) :=: τ_6,
Debug[Constr]: constr_3 : [worst,global,exact] IsLessEqual (τ_3,τ_0),
Debug[Constr]: constr_7 : [global,exact] IsSupremum (τ_4,τ_1) :=: τ_3,
Debug[Constr]: constr_0 : [worst,global,exact] IsLessEqual (τ_0,τ_1)
Debug[Constr]:    - others:
Debug[Constr]: []
Debug[Subtyping]: ~~~~ collapsing cycles of (τ_4,τ_3)
Debug[Subtyping]: ~~~~ graph is fromList [(τ_5,[τ_6]),(τ_3,[τ_0]),(τ_0,[τ_6,τ_1]),(τ_1,[τ_3]),(τ_4,[τ_3])]
Debug[Subtyping]: ~~~~ found cycles [] unifying with τ_3
Debug[Subtyping]: ~~~~ collapsing cycles of (τ_1,τ_3)
Debug[Subtyping]: ~~~~ graph is fromList [(τ_5,[τ_6]),(τ_3,[τ_0]),(τ_0,[τ_6,τ_1]),(τ_1,[τ_3]),(τ_4,[τ_3])]
Debug[Subtyping]: ~~~~ found cycles [] unifying with τ_3
```
which seems to not find an existing cycle `3 -> 0 -> 1 -> 3`.

**Comment by MxmUrw** [comment]
@ooovi , this does not seem to be a bug with my rewrite, but rather with the cycle solving, right?

**Comment by ooovi** [comment]
closed with 7dbb117
